django/db/models/fields/related_descriptors.py

```diff
@@ -912,6 +912,30 @@ def get_prefetch_queryset(self, instances, queryset=None):
                 False,
             )
 
+        def count(self):
+            # If the through's target field's foreign integrity is enforced
+            # the COUNT can be performed against the through table instead of
+            # INNER JOIN'ing the target table.
+            if self.target_field.db_constraint:
+                hints = {'instance': self.instance}
+                manager = self.through._base_manager.db_manager(using=self._db, hints=hints)
+                return manager.filter(
+                    **{self.source_field_name: self.instance.pk}
+                ).count()
+            return super().count()
+
+        def exists(self):
+            # If the through's target field's foreign integrity is enforced
+            # the EXISTS can be performed against the through table instead of
+            # INNER JOIN'ing the target table.
+            if self.target_field.db_constraint and not self.target_field.null:
```

Comment:
Even if it's not tested above `self.target_field.null` should be considered as well in `count()` and a test added.

By the way, the optimization is still doable when `self.target_field.null`, you just have to make sure to include `filters['%s__isnull' % self.target_field_name] = False` in the manager filters.